File: src/Array/UnionAndIntersection.java
// https://practice.geeksforgeeks.org/problems/union-of-two-arrays/0
// Reference Links :-
// https://www.geeksforgeeks.org/find-union-and-intersection-of-two-unsorted-arrays/
// https://www.geeksforgeeks.org/union-and-intersection-of-two-sorted-arrays-2/

package Array;

// Given two arrays a[] and b[] of size n and m respectively. The task is to find union between these two arrays.
// Union of the two arrays can be defined as the set containing distinct elements from both the arrays.

// IMPORTANT: If there are repetitions, then only one occurrence of element should be printed in the union/intersection.

import java.util.Arrays;
import java.util.HashSet;

public class UnionAndIntersection {

    // UNION
    // Approach 1 - Modified merge method
    // This method assumes that the arrays are sorted
    // If the arrays are not sorted then use Quicksort to sort the arrays then apply the doUnionOnSortedArrays method
    // Union array generated in this approach is automatically sorted

    // Time Complexity (Sorted Arrays as Input): O(M + N) where M & N are the lengths of the two arrays
    // Space Complexity: O(M + N)
    // ---------------------------
    // Time Complexity (Unsorted Arrays as Input): O(MlogM + NlogN) (Quicksort)
    // Space Complexity: O(M + N)

    public static int[] doUnionOnSortedArrays(int[] arr1, int[] arr2) {
        // Max size of union array will be when all elements are distinct in both arrays
        int[] unionArr = new int[arr1.length + arr2.length];
        int arr1Pointer = 0, arr2Pointer = 0, unionArrPointer = 0;

        while (arr1Pointer < arr1.length && arr2Pointer < arr2.length) {
            if (arr1[arr1Pointer] < arr2[arr2Pointer]) {
                // Copy element from arr1 to union array only if it is not already present
                if (unionArrPointer == 0 || unionArr[unionArrPointer - 1] != arr1[arr1Pointer]) {
                    unionArr[unionArrPointer++] = arr1[arr1Pointer++];
                } else {
                    arr1Pointer++;
                }
            } else if (arr1[arr1Pointer] > arr2[arr2Pointer]) {
                // Copy element from arr2 to union array only if it is not already present
                if (unionArrPointer == 0 || unionArr[unionArrPointer - 1] != arr2[arr2Pointer]) {
                    unionArr[unionArrPointer++] = arr2[arr2Pointer++];
                } else {
                    arr2Pointer++;
                }
            } else if (arr1[arr1Pointer] == arr2[arr2Pointer]) {
                // Both elements are equal, insert one to the union array and increment both pointers
                if (unionArrPointer == 0 || unionArr[unionArrPointer - 1] != arr1[arr1Pointer]) {
                    unionArr[unionArrPointer++] = arr1[arr1Pointer++];
                } else {
                    arr1Pointer++;
                }
                arr2Pointer++;
            }
        }

        // Copy remaining elements if any (happens when one array has greater length than the other)
        while (arr1Pointer < arr1.length) {
            if (unionArrPointer == 0 || unionArr[unionArrPointer - 1] != arr1[arr1Pointer]) {
                unionArr[unionArrPointer++] = arr1[arr1Pointer++];
            } else {
                arr1Pointer++;
            }
        }
        while (arr2Pointer < arr2.length) {
            if (unionArrPointer == 0 || unionArr[unionArrPointer - 1] != arr2[arr2Pointer]) {
                unionArr[unionArrPointer++] = arr2[arr2Pointer++];
            } else {
                arr2Pointer++;
            }
        }

        // Remove the extra space from the array
        unionArr = Arrays.copyOf(unionArr, unionArrPointer);
        return unionArr;
    }

    // ---------------------------------------------------------------------------------------------------------------//

    // UNION
    // Approach 2 - HashSet (Collection Framework)
    // Set is a data structure that allows only the distinct elements in it.
    // So, when we put the elements of both the array into the set we will get only the distinct elements
    // Which is basically the union operation over the arrays.

    // Iterate over both arrays from 0 to min(M, N) - 1 and add both all the elements to the set
    // Iterate over the larger array and add remaining elements to the set

    // This works on unsorted as well as sorted arrays
    // However, the final union array is not sorted.

    // Time Complexity: O(max(M, N))
    // Space Complexity: O(M + N)

    public static int[] doUnionUsingSet(int[] arr1, int[] arr2) {
        int M = arr1.length;
        int N = arr2.length;
        int min = Math.min(N, M);

        HashSet<Integer> set = new HashSet<>();
        for (int i = 0; i < min; i++) {
            set.add(arr1[i]);
            set.add(arr2[i]);
        }

        if (min == M) {
            for (int i = M; i < N; i++) {
                set.add(arr2[i]);
            }
        } else {
            for (int i = N; i < M; i++) {
                set.add(arr1[i]);
            }
        }

        return set.stream().mapToInt(i -> i).toArray();
    }

    // ---------------------------------------------------------------------------------------------------------------//

    // INTERSECTION
    // Approach 1 - Modified merge method
    // This method assumes that the arrays are sorted
    // If the arrays are not sorted then use Quicksort to sort the arrays then apply the doIntersectionOnSortedArrays method
    // Intersection array generated in this approach is automatically sorted

    // Time Complexity (Sorted Arrays as Input): O(M + N) where M & N are the lengths of the two arrays
    // Space Complexity: O(M + N)
    // ---------------------------
    // Time Complexity (Unsorted Arrays as Input): O(MlogM + NlogN) (Quicksort)
    // Space Complexity: O(M + N)

    public static int[] doIntersectionOnSortedArrays(int[] arr1, int[] arr2) {
        // Max size of intersection array will be when all elements are distinct in both arrays
        int[] intersectionArr = new int[arr1.length + arr2.length];
        int arr1Pointer = 0, arr2Pointer = 0, intersectionArrPointer = 0;

        while (arr1Pointer < arr1.length && arr2Pointer < arr2.length) {
            if (arr1[arr1Pointer] < arr2[arr2Pointer]) {
                arr1Pointer++;
            } else if (arr1[arr1Pointer] > arr2[arr2Pointer]) {
                arr2Pointer++;
            } else if (arr1[arr1Pointer] == arr2[arr2Pointer]) {
                // Both elements are equal, insert one to the intersection array and increment both pointers
                if (intersectionArrPointer == 0 || intersectionArr[intersectionArrPointer - 1] != arr1[arr1Pointer]) {
                    intersectionArr[intersectionArrPointer++] = arr1[arr1Pointer++];
                } else {
                    arr1Pointer++;
                }
                arr2Pointer++;
            }
        }

        // Remove the extra space from the array
        intersectionArr = Arrays.copyOf(intersectionArr, intersectionArrPointer);
        return intersectionArr;
    }

    // ---------------------------------------------------------------------------------------------------------------//

    // INTERSECTION
    // Approach 2 - HashSet (Collection Framework)
    // Put all elements of first array in the set
    // Iterate over the elements of the second array, if present in the set then add them to the intersection array

    // This works on unsorted as well as sorted arrays
    // However, the final intersection array is not sorted.

    // Time Complexity: O(M + N) (Assuming that hash table insert & search take O(1))
    // Space Complexity: O(M + N)

    public static int[] doIntersectionUsingSet(int[] arr1, int[] arr2) {
        int M = arr1.length;
        int N = arr2.length;
        int min = Math.min(M, N);

        HashSet<Integer> set = new HashSet<>();
        int[] intersectionArr = new int[arr1.length + arr2.length];
        int nextFreePosition = 0;

        for (int item : (min == M ? arr1 : arr2)) {
            set.add(item);
        }
        for (int item : (min == M ? arr2 : arr1)) {
            if (set.contains(item)) {
                intersectionArr[nextFreePosition++] = item;
            }
        }

        // Remove extra space
        intersectionArr = Arrays.copyOf(intersectionArr, nextFreePosition);

        // Remove duplicates
        HashSet<Integer> intersectionSet = new HashSet<>();
        for (int item : intersectionArr) {
            intersectionSet.add(item);
        }

        return intersectionSet.stream().mapToInt(i -> i).toArray();
    }

    public static void main(String[] args) {
        // UNION
        // If arrays are unsorted, sort them using quicksort before using doUnionOnSortedArrays
        // Final union array is automatically sorted
        int[] arr1 = {1, 6, 25, 32, 54, 85, 85};
        int[] arr2 = {2, 85, 85};
        System.out.println(Arrays.toString(doUnionOnSortedArrays(arr1, arr2))); // [1, 2, 6, 25, 32, 54, 85]

        // UNION
        // Works on unsorted as well as sorted arrays
        // However final union array is not sorted in this approach
        int[] arr3 = {6, 1, 85, 32, 25, 85, 54};
        int[] arr4 = {85, 2, 85};
        System.out.println(Arrays.toString(doUnionUsingSet(arr3, arr4))); // [32, 1, 2, 85, 6, 54, 25]


        // INTERSECTION
        // If arrays are unsorted, sort them using quicksort before using doIntersectionOnSortedArrays
        // Final intersection array is automatically sorted
        int[] arr5 = {1, 6, 25, 32, 54, 85, 85};
        int[] arr6 = {2, 85, 85};
        System.out.println(Arrays.toString(doIntersectionOnSortedArrays(arr5, arr6))); // [85]

        // INTERSECTION
        // Works on unsorted as well as sorted arrays
        // However final intersection array is not sorted in this approach
        int[] arr7 = {6, 1, 85, 32, 25, 85, 54};
        int[] arr8 = {85, 2, 85};
        System.out.println(Arrays.toString(doIntersectionUsingSet(arr7, arr8))); // [85]
    }
}

File: Love Babbar 450 DSA Questions/src/Cpp/Arrays/UnionAndIntersection.cpp
// https://practice.geeksforgeeks.org/problems/union-of-two-arrays3538/1

#include <iostream>
#include <vector>
#include <unordered_set>
#include <algorithm>
#include <utility> // For pair

// Given two arrays a[] and b[] of size n and m respectively.
// The task is to find union between these two arrays.
// Bonus: Write code to find the intersection b/w 2 arrays.

using namespace std;
void printVector(vector<int> v);
void printSet(unordered_set<int> s);

// UNION
// Approach 1 - Sort the arrays, then merge.
// Final union array is also sorted
// Time Complexity: O(NlogN + MlogM) <- sort() function
// Space Complexity: O(logN + logM) <- sort() function
vector<int> findUnionUsingVector(int arr1[], int n1, int arr2[], int n2)
{
    // Sort the two arrays if they are not sorted
    sort(arr1, arr1 + n1);
    sort(arr2, arr2 + n2);
    vector<int> unionArr;
    int arr1P{0};
    int arr2P{0};
    while (arr1P < n1 && arr2P < n2)
    {
        if (arr1[arr1P] == arr2[arr2P])
        {
            if (count(unionArr.begin(), unionArr.end(), arr1[arr1P]) == 0)
            {
                unionArr.push_back(arr1[arr1P]);
            }
            arr1P++;
            arr2P++;
        }
        else if (arr1[arr1P] < arr2[arr2P])
        {
            if (count(unionArr.begin(), unionArr.end(), arr1[arr1P]) == 0)
            {
                unionArr.push_back(arr1[arr1P]);
            }
            arr1P++;
        }
        else
        {
            if (count(unionArr.begin(), unionArr.end(), arr2[arr2P]) == 0)
            {
                unionArr.push_back(arr2[arr2P]);
            }
            arr2P++;
        }
    }
    while (arr1P < n1)
    {
        unionArr.push_back(arr1[arr1P++]);
    }
    while (arr2P < n2)
    {
        unionArr.push_back(arr2[arr2P++]);
    }

    return unionArr;
}

// Approach 2 - Use unordered_set
// Final union array is not sorted since we are using unordered_set
// If sorted array is needed, sort it using the sort function or use set
// Time Complexity: O(N + M)  <- Assuming no collisions so insertion is O(1) always
// Space Complexity: O(N + M)
unordered_set<int> findUnionUsingSet(int arr1[], int n1, int arr2[], int n2)
{
    unordered_set<int> unionSet;
    for (int i = 0; i < n1; i++)
    {
        unionSet.insert(arr1[i]);
    }
    for (int i = 0; i < n2; i++)
    {
        unionSet.insert(arr2[i]);
    }

    return unionSet;
}

// INTERSECTION
// Approach 1 - Sort the arrays, then use modified merge method
// Final intersection array is also sorted.
// Time complexity: O(NlogN + MlogM)
// Space complexity: O(logN)
vector<int> findIntersectionUsingVector(int arr1[], int n1, int arr2[], int n2)
{
    sort(arr1, arr1 + n1);
    sort(arr2, arr2 + n2);
    vector<int> interVec;
    int arr1P{0};
    int arr2P{0};
    while (arr1P < n1 && arr2P < n2)
    {
        if (arr1[arr1P] == arr2[arr2P])
        {
            if (!count(interVec.begin(), interVec.end(), arr1[arr1P]))
            {
                interVec.push_back(arr1[arr1P]);
            }
            arr1P++;
            arr2P++;
        }
        else if (arr1[arr1P] < arr2[arr2P])
        {
            arr1P++;
        }
        else
        {
            arr2P++;
        }
    }

    return interVec;
}

// Approach 2 - Use unordered_set
// Final intersection array is not sorted since we are using unordered_set
// If sorted array is needed, sort it using the sort function or use set
// Time Complexity: O(N + M)  <- Assuming no collisions so insertion is O(1) always
// Space Complexity: O(N + M)
unordered_set<int> findIntersectionUsingSet(int arr1[], int n1, int arr2[], int n2)
{
    int *smallerArr;
    int sLen;
    int *largerArr;
    int lLen;
    if (n1 < n2)
    {
        smallerArr = arr1;
        sLen = n1;
        largerArr = arr2;
        lLen = n2;
    }
    else
    {
        smallerArr = arr2;
        sLen = n2;
        largerArr = arr1;
        lLen = n1;
    }

    unordered_set<int> s;
    // Insert all elements of smallerArr in s
    for (int i = 0; i < sLen; i++)
    {
        s.insert(*(smallerArr + i));
    }

    unordered_set<int> intersectionSet;
    for (int i = 0; i < lLen; i++)
    {
        if (s.count(*(largerArr + i)) != 0)
        {
            intersectionSet.insert(*(largerArr + i));
        }
    }
    return intersectionSet;
}

int main()
{
    // Create arrays for testing the functions
    int arr1[]{85, 6, 1, 32, 54, 25, 85};
    const int n1 = sizeof(arr1) / sizeof(arr1[0]);
    int arr2[]{85, 2, 85};
    const int n2 = sizeof(arr2) / sizeof(arr2[0]);
    int arr3[n1];
    copy(arr1, arr1 + n1, arr3);
    int arr4[n2];
    copy(arr2, arr2 + n2, arr4);
    int arr5[n1];
    copy(arr1, arr1 + n1, arr5);
    int arr6[n2];
    copy(arr2, arr2 + n2, arr6);
    int arr7[n1];
    copy(arr1, arr1 + n1, arr7);
    int arr8[n2];
    copy(arr2, arr2 + n2, arr8);

    // Union
    vector<int> u1Vector{findUnionUsingVector(arr1, n1, arr2, n2)};
    printVector(u1Vector); // 1 2 6 25 32 54 85
    unordered_set<int> u2Set{findUnionUsingSet(arr3, n1, arr4, n2)};
    printSet(u2Set); // 85 2 54 25 32 6 1

    // Intersection
    vector<int> i1Vector{findIntersectionUsingVector(arr5, n1, arr6, n2)};
    printVector(i1Vector); // 85
    unordered_set<int> i2Set{findIntersectionUsingSet(arr7, n1, arr8, n2)};
    printSet(i2Set); // 85
    return 0;
}

void printVector(vector<int> v)
{
    for (int i : v)
    {
        cout << i << ' ';
    }
    cout << '\n';
}

void printSet(unordered_set<int> s)
{
    for (int i : s)
    {
        cout << i << ' ';
    }
    cout << '\n';
}